string query
string filter
---
device[] data
